### PR TITLE
[quant] Add QuantizedLSTM class

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79958
* #79957
* #79956
* __->__ #71189

The nn.LSTM is quantized through the custom module mechanism, which uses the nn.quantizable.LSTM for both observed and quantized paths. This is potentially a source of confusion. This creates a `quantized.LSTM` class, which completely takes the quantized path. Note that after this, the old usage will throw an error.

New way of using it:

```
>>> custom_module_config = {
...     'float_to_observed_custom_module_class': {
...         nn.LSTM: nn.quantizable.LSTM,
...     },
...     'observed_to_quantized_custom_module_class': {
...         nn.quantizable.LSTM: nn.quantized.LSTM,
...     }
... }
>>> tq.prepare(model, prepare_custom_module_class=custom_module_config)
>>> tq.convert(model, convert_custom_module_class=custom_module_config)
```

### CODE REVIEW

File: test/test_module_init.py [hunk 1]
```diff
@@ -366,6 +366,7 @@ def generate_tests(test_cls, constructor_arg_db):
         # See https://github.com/pytorch/pytorch/issues/55396
         torch.nn.quantized.Embedding,
         torch.nn.quantized.EmbeddingBag,
+        torch.nn.quantized.modules.rnn.LSTM,
```
Comment (jerryzh168) [inline test/test_module_init.py:369]: nit: torch.nn.quantized.LSTM (if it doesn't work we should import the module in https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/__init__.py

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

File: test/test_module_init.py (hunk 1 above)
Comment (vkuzo) [reply test/test_module_init.py:373]: this should be fixed before landing

File: torch/nn/quantizable/modules/activation.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def __init__(self, embed_dim: int, num_heads: int,
         self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=bias, **factory_kwargs)  # type: ignore[assignment]
 
         # Functionals
-        self.q_scaling_product = nnq.FloatFunctional()
```
Comment (vkuzo) [inline torch/nn/quantizable/modules/activation.py:76]: what's the reasoning for this change?

**Review by vkuzo — APPROVED** [review]

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

File: torch/nn/quantizable/modules/activation.py (hunk 2 above)
Comment (HDCharles) [reply torch/nn/quantizable/modules/activation.py:76]: NOPE. Without this change there's a circular import that gets created by the PR. Not sure if there's a better way to address this though...

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33451338).

**Pull request closed without merging**